src/core_tests/workers.rs

```diff
@@ -138,3 +145,114 @@ async fn pending_activities_only_returned_for_their_queue() {
+    assert_matches!(
+        core.poll_workflow_task(TEST_Q).await.unwrap_err(),
+        PollWfError::NoWorkerForQueue(_)
+    );
+}
+
+#[tokio::test]
+async fn after_shutdown_of_worker_get_shutdown_err() {
+    let t = canned_histories::single_timer("fake_timer");
+    let core = build_fake_core("fake_wf_id", t, &[1]);
+    let res = core.inner.poll_workflow_task(TEST_Q).await.unwrap();
+    assert_eq!(res.jobs.len(), 1);
+    core.inner.shutdown_worker(TEST_Q).await;
+    assert_matches!(
+        core.inner.poll_workflow_task(TEST_Q).await.unwrap_err(),
+        PollWfError::ShutDown
+    );
+}
+
+#[tokio::test]
+async fn after_shutdown_of_worker_can_be_reregistered() {
+    let t = canned_histories::single_timer("fake_timer");
+    let core = build_fake_core("fake_wf_id", t, &[1]);
+    let res = core.inner.poll_workflow_task(TEST_Q).await.unwrap();
+    assert_eq!(res.jobs.len(), 1);
+    core.inner.shutdown_worker(TEST_Q).await;
+    core.inner
+        .register_worker(
+            WorkerConfigBuilder::default()
+                .task_queue(TEST_Q)
+                .build()
+                .unwrap(),
+        )
+        .await
+        .unwrap();
+}
+
+#[tokio::test]
+async fn shutdown_worker_can_complete_pending_activation() {
+    let t = canned_histories::single_timer("fake_timer");
+    let core = build_fake_core("fake_wf_id", t, &[2]);
+    let res = core.inner.poll_workflow_task(TEST_Q).await.unwrap();
+    assert_eq!(res.jobs.len(), 1);
+    // Complete the timer, will queue PA
+    core.inner
+        .complete_workflow_task(WfActivationCompletion::from_cmds(
+            vec![StartTimer {
+                timer_id: "fake_timer".to_string(),
+                ..Default::default()
+            }
+            .into()],
+            res.run_id,
+        ))
+        .await
+        .unwrap();
+
+    core.inner.shutdown_worker(TEST_Q).await;
+    let res = core.inner.poll_workflow_task(TEST_Q).await.unwrap();
+    // The timer fires
+    assert_eq!(res.jobs.len(), 1);
+    core.inner
+        .complete_workflow_task(WfActivationCompletion::from_cmds(
+            vec![CompleteWorkflowExecution::default().into()],
+            res.run_id,
+        ))
+        .await
+        .unwrap();
+    // Since non-sticky, one more activation for eviction
+    core.inner.poll_workflow_task(TEST_Q).await.unwrap();
+    // Now it's shut down
+    assert_matches!(
+        core.inner.poll_workflow_task(TEST_Q).await.unwrap_err(),
+        PollWfError::ShutDown
+    );
+}
+
+#[tokio::test]
+async fn worker_shutdown_during_poll_doesnt_deadlock() {
+    let mut mock_gateway = MockManualGateway::new();
+    mock_gateway
+        .expect_poll_workflow_task()
+        .returning(move |_| {
+            // Slow poll response
+            async move {
+                let t = canned_histories::single_timer("fake_timer");
+                sleep(Duration::from_secs(1)).await;
```

Comment:
I usually would rather avoid sleeping in tests, they tend to break when CI server is under load and make you wait unnecessarily.
Would be much better to use some sort of signaling.